### PR TITLE
Avoid using feature flags and env variable to set the same parameter pt.1 emulation_mode

### PR DESCRIPTION
**Pull request by Marcondiro**
This PR follows discussion https://github.com/AFLplusplus/LibAFL/discussions/2505
**TLDR:** Using only the feature flag simplifies things a bit and allows the usage of optional dependencies

### CODE REVIEW

**Comment by Marcondiro** [comment]
Ubuntu-clippy CI fails because it is run with `--all-features`.
Upstream doesn't fail because it silently defaults to `usermode`, clippy is not run for `systemmode`

**Comment by domenukk** [comment]
--all-features should enabled a clippy feature which makes it work magically. Maybe just build usermode always?

**Comment by Marcondiro** [comment]
@domenukk I'm not sure I fully understood your message :smile:
How about avoiding `--all-features` for `libafl_qemu`?
I don't like this solution very much and maybe in the future the incompatible features thing could be reconsidered

**Comment by domenukk** [comment]
> @domenukk I'm not sure I fully understood your message 😄 How about avoiding `--all-features` for `libafl_qemu`? I don't like this solution very much and maybe in the future the incompatible features thing could be reconsidered

running clippy --all-features (IMHO) should just work(tm)

**Comment by Marcondiro** [comment]
The thing is that in `libafl_qemu` some features are incompatible (`usermode` and `systemmode` for instance), running with `--all-features` should fail.

So far the build script decided which feature to actually use and which silently ignore when running with `--all-features`.
In my opinion  having `--all-features` just working but actually not enabling all features is not ideal. Looking at the clippy script it looks like every feature of libafl_qemu is checked with clippy while it is not.

I thought this change would simplify and clean things a bit (no hidden defaults, fallback env vars, just plain feature), but if it looks more convoluted to you I'll just drop this PR no problem :)

**Comment by domenukk** [comment]
Features should be additive (see: https://doc.rust-lang.org/nightly/cargo/reference/features.html#feature-unification ) so I think we should either

- specify a default for what happens when both features are set
or
- take away the `usermode` feature and make it the default. Note that this is likely not possible for architectures, so...

I'm not sure what other crates do that have mutual exclusive dependencies, like crypto libs with backends etc?

**Comment by domenukk** [comment]
The link you posted literally states:
```
Instead of using mutually exclusive features, consider some other options:
...
* When there is a conflict, [choose one feature over another](https://doc.rust-lang.org/nightly/cargo/reference/features-examples.html#feature-precedence). The [cfg-if](https://crates.io/crates/cfg-if) package can help with writing more complex cfg expressions.
```

In this case it's easy - we default to usermode and x86_64, that's what we've done so far

**Comment by domenukk** [comment]
BTW I don't care either way, @rmalmain and/or @andreafioraldi should say how they want it

**Comment by andreafioraldi** [comment]
if there is a conflict i would fail with an error, as we already do for x86+big endian for instance

**Comment by domenukk** [comment]
Fine by me - but maybe if we still have the `clippy` feature that makes things "just work" `--all` will still do the right thing here?

**Comment by rmalmain** [comment]
we can also adapt clippy flags in the clippy script is the CI variable is on.
i don't think it makes sense to change this kind of general behaviour only to have `clippy --all` working.  
i think we shouldn't use env variables to set / unset features in general, it is confusing otherwise.
btw @Marcondiro, i think you shouldn't use `unreachable!` but `panic!` instead with an error message, it should be more clear to understand the problem.

File: libafl_qemu/libafl_qemu_sys/build_linux.rs [hunk 1]
```diff
@@ -14,24 +14,27 @@ macro_rules! assert_unique_feature {
...
+        compile_error!(concat!("At least one of the following features must be enabled:", $(" ", $feature),+));
+    };
+}
+
 pub fn build() {
-    // Make sure that exactly one qemu mode is set
+    // Make sure that at most one qemu mode is set
     assert_unique_feature!("usermode", "systemmode");
+    // Make sure that at least one qemu mode is set
+    assert_at_least_one_feature!("usermode", "systemmode");
```
Comment (Marcondiro) [inline libafl_qemu/libafl_qemu_sys/build_linux.rs:29]: @rmalmain these two assert should make the code down actually unreachable and print a proper message, do you prefer a `panic` anyway?

Comment (rmalmain) [reply libafl_qemu/libafl_qemu_sys/build_linux.rs:29]: ok we can keep it that way, it's fine

Comment (domenukk) [reply libafl_qemu/libafl_qemu_sys/build_linux.rs:29]: maybe write the reason for the unreachable assumption in the unreachable macro (I think that's possible?)

Comment (Marcondiro) [reply libafl_qemu/libafl_qemu_sys/build_linux.rs:29]: @domenukk Cool, I didn't know :+1: done

**Comment by domenukk** [comment]
What's the status here? Needs a rebase and then needs to get reviewed/merged @rmalmain

File: scripts/clippy.sh [hunk 2]
```diff
@@ -32,6 +44,11 @@ if [[ "$OSTYPE" == "linux-gnu"* ]]; then
    )
 fi
 
+# Do not use --all-features for the following projects
+NO_ALL_FEATURES=(
```
Comment (domenukk) [inline scripts/clippy.sh:48]: I'm still not a big fan of this solution, IMHO the clippy feature / --all-features should just make CI work?

Comment (domenukk) [reply scripts/clippy.sh:48]: (I guess we discussed this before but I forgot / still don't understand the reasoning against it)

Comment (Marcondiro) [reply scripts/clippy.sh:48]: The reason was:
> The thing is that in `libafl_qemu` some features are incompatible (`usermode` and `systemmode` for instance), running with `--all-features` should fail.
> 
> So far the build script decided which feature to actually use and which silently ignore when running with `--all-features`. In my opinion having `--all-features` just working but actually not enabling all features is not ideal. Looking at the clippy script it looks like every feature of libafl_qemu is checked with clippy while it is not.

IMHO the default features should _just work™_, not the `--all-features` here.

Also, in practice, this change allows the usage of optional dependencies

Comment (Marcondiro) [reply scripts/clippy.sh:48]: 100% agree about the clippy feature tho

**Comment by Marcondiro** [comment]
@rmalmain should be ready for review

File: scripts/parallellize_cargo_check.py [hunk 3]
```diff
@@ -36,6 +36,20 @@
 ]:
     print("Running ", task)
     print(os.environ)
+    if (("--manifest-path libafl_qemu/Cargo.toml" in task or "--manifest-path libafl_qemu/libafl_qemu_sys/Cargo.toml" in task)
```
Comment (rmalmain) [inline scripts/parallellize_cargo_check.py:39]: not a huge fan of this one, it's not very readable imho

Comment (Marcondiro) [reply scripts/parallellize_cargo_check.py]: better now?

Comment (rmalmain) [reply scripts/parallellize_cargo_check.py]: yes, looks good

**Comment by rmalmain** [comment]
it's hard to tell if your pr is gonna break something at the moment, half of the CI is broken

**Pull request merged**